pkg/controllers/disruption/orchestration/queue.go

```diff
@@ -307,32 +305,26 @@ func (q *Queue) waitOrTerminate(ctx context.Context, cmd *Command) error {
 // Add adds commands to the Queue
 // Each command added to the queue should already be validated and ready for execution.
 func (q *Queue) Add(cmd *Command) error {
-	providerIDs := lo.Map(cmd.candidates, func(s *state.StateNode, _ int) string {
-		return s.ProviderID()
-	})
 	// First check if we can add the command.
-	if q.HasAny(providerIDs...) {
+	if q.HasAny(cmd.candidates...) {
 		return fmt.Errorf("candidate is being disrupted")
 	}
 
 	cmd.timeAdded = q.clock.Now()
-	q.mu.Lock()
-	for _, candidate := range cmd.candidates {
-		q.providerIDToCommand[candidate.ProviderID()] = cmd
-	}
-	q.mu.Unlock()
-	q.TypedRateLimitingInterface.Add(cmd)
+	q.Lock()
+	defer q.Unlock()
+	q.nodeClaimToCommand[client.ObjectKeyFromObject(cmd.KeyNodeClaim)] = cmd
 	return nil
 }
 
 // HasAny checks to see if the candidate is part of an currently executing command.
-func (q *Queue) HasAny(ids ...string) bool {
-	q.mu.RLock()
-	defer q.mu.RUnlock()
+func (q *Queue) HasAny(candidates ...*state.StateNode) bool {
+	q.RLock()
+	defer q.RUnlock()
 
 	// If the mapping has at least one of the candidates' providerIDs, return true.
-	_, ok := lo.Find(ids, func(id string) bool {
-		_, ok := q.providerIDToCommand[id]
+	_, ok := lo.Find(candidates, func(n *state.StateNode) bool {
+		_, ok := q.nodeClaimToCommand[client.ObjectKeyFromObject(n.NodeClaim)]
```

Comment:
We need to iterate through all of the candidates in the command, right? This is probably pretty expensive, we'd probably still want to maintain the providerID map

Reply:
IIUC the map doesn't have every candidate as a key - it just has a candidate which is chosen as the representative for the command. To check that any given candidate is in a command we could first check if it's a key in the map, but then we'd need to iterate through every nodeclaim in each command